src/WebJobs.Script/Binding/GeneralScriptBindingProvider.cs

```diff
@@ -70,28 +70,30 @@ public override bool TryResolveAssembly(string assemblyName, AssemblyLoadContext
         // Read the properties and convert that into a System.Type.
         internal static Type GetRequestedType(ScriptBindingContext context)
         {
-            Type type = ParseDataType(context);
+            Type type = ParseDataType(context, out bool typeSpecified);
 
             Cardinality cardinality;
             if (!Enum.TryParse<Cardinality>(context.Cardinality, true, out cardinality))
             {
                 cardinality = Cardinality.One; // default
             }
 
+            // arrays are supported for both trigger input as well
+            // as output bindings
             if (cardinality == Cardinality.Many)
             {
-                // arrays are supported for both trigger input as well
-                // as output bindings
-                type = type.MakeArrayType();
+                // Default to string array type
+                type = typeSpecified ? type.MakeArrayType() : typeof(string[]);
```

Comment:
Since the v1 version of this function didn't have this logic, and you're saying v1 logic was correct, I'm wondering if the issue isn't instead in the binding or worker code? Where is the actual problem?

Reply:
In v1 it might be that the conversion logic handling this was [here](https://github.com/Azure/azure-functions-host/blob/f41da7b5b7573f94ed06eef3c0b7d54b66df2ea3/src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs#L680)

Reply:
I think the logic was handled here: https://github.com/Azure/azure-functions-host/blob/v1.x/src/WebJobs.Script/Binding/ServiceBusScriptBindingProvider.cs/

If you have any context on where this logic may have been or should have been moved to, that'd be great :)

Reply:
Here's what I think we should do:

- since the whole concept of Default type is for supporting script languages, we need a default type that our worker channel can handle. Therefore we need to dictate it - not default to typeof(object) if no dataType specified, which allows the extension to dictate it
- Change ParseDataType to return a default of typeof(string) if no user specified dataType present. Assuming string is the default worker channel type supported?
- I think across the board, even when no explicit dataType is specified and requestedType defaults to typeof(object), GetDefaultType() ends up defaulting to a string type already anyways?

Separately we need to fix JobHostMetadataProvider so it accepts a null requestType and doesn't do all the typeof(object) conversions that are currently baked in. The typeof(object) special casing across all the implementations of GetDefaultType in the SDK is because this logic was originally copied from Functions so the above defaulting to typeof(object) was copied over. We should replace these with null checks.

Reply:
@mathewc 
Pragna brought this up in an email, but the [supported types](https://github.com/Azure/azure-functions-language-worker-protobuf/blob/dev/src/proto/FunctionRpc.proto#L260) for out-of-proc languages are:
```
message TypedData {
  oneof data {
    string string = 1;
    string json = 2;
    bytes bytes = 3;
    bytes stream = 4;
    RpcHttp http = 5;
    sint64 int = 6;
    double double = 7;
  }
}
```
TypedData of type json is selected to be JSON [in the host, in RpcMessageConversionExtensions.cs](https://github.com/Azure/azure-functions-host/blob/395b0c9929e8ac4066e0d7965aa6a1374cdbd019/src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs#L61-L64). So it looks like JObjects are supported, but there's an issue in the binding, where it tries to turn [JsonConvert.DeserializeObject](https://github.com/Azure/azure-webjobs-sdk/blob/254a9b6cdd07313f695941349739b9fa22c6a888/src/Microsoft.Azure.WebJobs.Host/Triggers/TriggerArgumentBinding/PocoTriggerArgumentBinding.cs#L47) the "object" type. 

I think in V1, we used to have DefaultType of EventHubs special-cased to be string. Then, in [TryConvertJson](https://github.com/Azure/azure-functions-host/blob/f41da7b5b7573f94ed06eef3c0b7d54b66df2ea3/src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs#L680) (the snippet you first linked), we would make the passed object JSON if possible.

Should we try doing something like that here too? Meaning we default to string, and then add TryConvertJson on string inputs in [RpcMessageConversionExtensions.cs](https://github.com/Azure/azure-functions-host/blob/395b0c9929e8ac4066e0d7965aa6a1374cdbd019/src/WebJobs.Script/Rpc/MessageExtensions/RpcMessageConversionExtensions.cs)?